integrators/embedded_explicit_runge_kutta_nyström_integrator_body.hpp

```diff
@@ -70,50 +70,49 @@ inline EmbeddedExplicitRungeKuttaNyströmIntegrator::
 
 template<typename Position>
 void EmbeddedExplicitRungeKuttaNyströmIntegrator::Solve(
-    RightHandSideComputation<Position> compute_acceleration,
-    SystemState<Position, Variation<Position>> const& initial_value,
-    Time const& t_final,
-    Time const& first_time_step,
-    StepSizeController<Position> step_size_controller,
-    double const safety_factor,
-    not_null<Solution<Position, Variation<Position>>*> const solution) const {
-  using Displacement = Difference<Position>;
-  using Velocity = Variation<Position>;
-  using Acceleration = Variation<Velocity>;
+    IntegrationProblem<ODE<Position>> const& problem,
+    AdaptiveStepSize<ODE<Position>> const& adaptive_step_size) const {
+  using Displacement = typename ODE<Position>::Displacement;
+  using Velocity = typename ODE<Position>::Velocity;
+  using Acceleration = typename ODE<Position>::Acceleration;
 
   // Argument checks.
-  int const dimension = initial_value.positions.size();
-  CHECK_EQ(initial_value.momenta.size(), dimension);
-  CHECK_NE(first_time_step, Time());
-  Sign const integration_direction = Sign(first_time_step);
+  CHECK_NOTNULL(problem.initial_state);
+  int const dimension = problem.initial_state->positions.size();
+  CHECK_EQ(problem.initial_state->velocities.size(), dimension);
+  CHECK_NE(adaptive_step_size.first_time_step, Time());
+  Sign const integration_direction =
+      Sign(adaptive_step_size.first_time_step);
   if (integration_direction.Positive()) {
     // Integrating forward.
-    CHECK_LT(initial_value.time.value, t_final);
+    CHECK_LT(problem.initial_state->time.value, problem.t_final);
   } else {
     // Integrating backward.
-    CHECK_GT(initial_value.time.value, t_final);
+    CHECK_GT(problem.initial_state->time.value, problem.t_final);
   }
-  CHECK_GT(safety_factor, 0);
-  CHECK_LT(safety_factor, 1);
+  CHECK_GT(adaptive_step_size.safety_factor, 0);
+  CHECK_LT(adaptive_step_size.safety_factor, 1);
+
+  typename ODE<Position>::SystemState current_state = *problem.initial_state;
 
   // Time step.
-  Time h = first_time_step;
+  Time h = adaptive_step_size.first_time_step;
   // Current time.
-  DoublePrecision<Time> t = initial_value.time;
+  DoublePrecision<Instant>& t = current_state.time;
 
   // Position increment (high-order).
   std::vector<Displacement> ∆q_hat(dimension);
   // Velocity increment (high-order).
   std::vector<Velocity> ∆v_hat(dimension);
   // Current position.
-  std::vector<DoublePrecision<Position>> q_hat = initial_value.positions;
+  std::vector<DoublePrecision<Position>>& q_hat = current_state.positions;
```

Comment:
Non-const references are fine but unusual.  Comment that this is really what you want.